src/operator/batch_norm-inl.h

```diff
@@ -50,20 +52,294 @@ struct BatchNormParam : public dmlc::Parameter<BatchNormParam> {
+    const size_t channels = dim > 1 ? shape[1] : 1;
+    const size_t matrixSize = shape.Size() / (channels * num);
+
+    for (size_t batchItem = 0; batchItem < num; ++batchItem) {
+      #pragma openmp for
+      for (size_t channel = 0; channel < channels; ++channel) {
+        for (size_t i = 0; i < matrixSize; ++i) {
+          onData(channel, in_data++);
+        }
+      }
+    }
+  }
+
+  /*! \brief Fast-foreach when you don't care about the position other than channel */
+  template<typename Shape, typename OnData>
+  static inline void forEachFast(Dtype *in_data, const Shape& shape,
+                                 const size_t channel, OnData onData) {
+    const size_t dim = shape.ndim();
+    const size_t num = shape[0];
+    const size_t channels = dim > 1 ? shape[1] : 1;
+    const size_t matrixSize = shape.Size() / (channels * num);
+
+    size_t indices[dim] = {0, };
+    indices[1] = channel;
+
+    for (size_t batchItem = 0; batchItem < num; ++batchItem) {
+      indices[0] = batchItem;
+      Dtype *data = in_data + offset(shape, indices, dim);
+      for (size_t i = 0; i < matrixSize; ++i) {
+        onData(data++);
+      }
+    }
+  }
+
+  /*! \brief Fast-foreach when you don't care about the position other than channel */
+  template<typename Shape, typename OnData>
+  static inline void forEachFast(const Dtype *in_data, Dtype *out_data,
+                                 const Shape& shape, const size_t channel, OnData onData) {
+    const size_t dim = shape.ndim();
+    const size_t num = shape[0];
+    const size_t channels = dim > 1 ? shape[1] : 1;
+    const size_t matrixSize = shape.Size() / (channels * num);
+
+    size_t indices[dim] = {0, };
+    indices[1] = channel;
+
+    for (size_t batchItem = 0; batchItem < num; ++batchItem) {
+      indices[0] = batchItem;
+      const size_t off = offset(shape, indices, dim);
+      const Dtype *data = in_data + off;
+      Dtype *odata = out_data + off;
+      for (size_t i = 0; i < matrixSize; ++i) {
+        onData(data++, odata++);
+      }
+    }
+  }
+
+  /*! \brief Fast-foreach when you don't care about the position other than channel */
+  template<typename Shape, typename OnData>
+  static inline void forEachFast(const Dtype *in_data, Dtype *out_data,
+                                 const Shape& shape, OnData onData) {
+    const size_t dim = shape.ndim();
+    const size_t num = shape[0];
+    const size_t channels = dim > 1 ? shape[1] : 1;
+    const size_t matrixSize = shape.Size() / (channels * num);
+
+    for (size_t batchItem = 0; batchItem < num; ++batchItem) {
+      #pragma omp parallel for
+      for (size_t channel = 0; channel < channels; ++channel) {
+        for (size_t i = 0; i < matrixSize; ++i) {
+          onData(channel, in_data++, out_data++);
+        }
+      }
+    }
+  }
+
+  /*! \brief Compute the mean of each input channel */
+  template<typename Shape>
+  static inline void computeMean(const Dtype *in_data,
```

Comment:
Seems you implemented computing mean and variance without using gemm that has been optmized internally. According to your benchmark results, new CPU is slower than the old CPU. Any good reasons?